certbot/certbot/interfaces.py

```diff
@@ -40,8 +48,192 @@ def save(self, account, client):  # pragma: no cover
+        """Keys storage."""
+
+    @property
+    @abstractmethod
+    def temp_checkpoint_dir(self) -> str:
+        """Temporary checkpoint directory."""
+
+    @property
+    @abstractmethod
+    def no_verify_ssl(self) -> bool:
+        """Disable verification of the ACME server's certificate."""
+
+    @property
+    @abstractmethod
+    def http01_port(self) -> int:
+        """Port used in the http-01 challenge.
+
+        This only affects the port Certbot listens on.
+        A conforming ACME server will still attempt to connect on port 80.
+        """
+
+    @property
+    @abstractmethod
+    def http01_address(self) -> str:
+        """The address the server listens to during http-01 challenge."""
+
+    @property
+    @abstractmethod
+    def https_port(self) -> int:
+        """Port used to serve HTTPS.
+
+        This affects which port Nginx will listen on after a LE certificate
+        is installed.
+        """
+
+    @property
+    @abstractmethod
+    def pref_challs(self) -> List[str]:
+        """List of user specified preferred challenges.
+
+        Sorted with the most preferred challenge listed first.
+        """
+
+    @property
+    @abstractmethod
+    def allow_subset_of_names(self) -> bool:
+        """Allow only a subset of names to be authorized to perform validations.
+
+        When performing domain validation, do not consider it a failure
+        if authorizations can not be obtained for a strict subset of
+        the requested domains. This may be useful for allowing renewals for
+        multiple domains to succeed even if some domains no longer point
+        at this system.
+        """
+
+    @property
+    @abstractmethod
+    def strict_permissions(self) -> bool:
+        """Enable strict permissions checks.
+
+        Require that all configuration files are owned by the current
+        user; only needed if your config is somewhere unsafe like /tmp/.
+        """
+
+    @property
+    @abstractmethod
+    def disable_renew_updates(self) -> bool:
+        """Disable renewal updates.
+
+        If updates provided by installer enhancements when Certbot is being run
+        with \"renew\" verb should be disabled.
+        """
+
+    @property
+    @abstractmethod
+    def preferred_chain(self) -> str:
+        """Set the preferred certificate chain to issue a certificate.
+
+        If the CA offers multiple certificate chains, prefer the chain whose
+        topmost certificate was issued from this Subject Common Name.
+        If no match, the default offered chain will be used.
+        """
+
+
+class IPluginFactory(zope.interface.Interface):  # pylint: disable=inherit-non-class
+    """Deprecated, use certbot.interfaces.Plugin as ABC instead."""
+
+
+class IPlugin(zope.interface.Interface):  # pylint: disable=inherit-non-class
+    """Deprecated, use certbot.interfaces.Plugin as ABC instead."""
+
+
+@zope.interface.implementer(IPlugin)
+@zope.interface.provider(IPluginFactory)
```

Comment:
What's the reason for using `implementer` or `provider` on the new classes defined here? Just for the sake of stability and avoiding bugs, I think it'd be kind of nice if the new classes never had anything to do with `zope`.

Reply:
The only reason is to continue to have `zope.interface` saying that `IPlugin` and `IPluginFactory` are still implemented/provided for the existing plugin classes that are using one of the interfaces or one of the abstract classes like `certbot.plugins.common.Installer` or `certbot.plugins.dns_common.DNSAuthenticator`.

Since these two classes are largely used in the third party community plugin, I felt that I could cheaply evacuate one of the potential compatibility breakage with this move.

Reply:
Honestly, after reviewing several typical implementation in third parties, I think that we do not care about adding the decorator to these new classes. The zope interfaces are never used directly, in favor of the ABC classes defined in `certbot.plugins.common`. For this only known use case, everything will work out-of-the-box for these plugins without any modification.

This implementation was mainly extra care. Let's remove it.